ext/extralite/database.c

```diff
@@ -386,6 +386,34 @@ VALUE Database_prepare(VALUE self, VALUE sql) {
   return rb_funcall(cPreparedStatement, ID_NEW, 2, self, sql);
 }
 
+/* call-seq:
+ *   db.interrupt
+ *
+ * This function causes any pending database operation to abort
+ * and return at its earliest opportunity.
+ *
+ * It is safe to call this routine from a thread different from
+ * the thread that is currently running the database operation.
+ * But it is not safe to call this routine with a database connection
+ * that is closed or might close before sqlite3_interrupt() returns.
+ *
+ * If an SQL operation is very nearly finished at the time
+ * when sqlite3_interrupt() is called, then it might not have
+ * an opportunity to be interrupted and might continue to completion.
+ *
+ * An SQL operation that is interrupted will return SQLITE_INTERRUPT.
+ * If the interrupted SQL operation is an INSERT, UPDATE, or DELETE
+ * that is inside an explicit transaction, then the entire transaction
+ * will be rolled back automatically.
+ */
```

Comment:
I don't really see the point in copying the text from https://sqlite.org/c3ref/interrupt.html. It deals with the C API and is too verbose. Rather something along the lines of:

Interrupts a long running query. This method is to be called from a different thread than the one running the query. Upon calling `#interrupt` the running query will stop and raise a `Extralite::InterruptError` exception.

Reply:
I am a bit worried about the possible race condition in between the `db#interrupt` and `safe_query_hash/stmt_iterate` loop. It's actually not easy to use. But implementing the proper interruption support would be not a very trivial task in the sense that it will complicate the code.

The proper implementation would require an atomically controlled interruption flag in `Database_t` and another flag that the query is executing (by safe_query_hash) and actually a mutex to control both atomically. But having just db#interrupt as it is is better than nothing and similar `Future`-like objects can be implemented on the app level with simple spins in the loop in timeout threads.

Reply:
I don't see where how a race condition could arise. If an exception is raised, it would necessarily stop the `stmt_iterate` loop. Let me know if I'm missing something, but it does not look to me that this is an issue.

The `sqlite3_interrupt` [docs](https://sqlite.org/c3ref/interrupt.html) do mention that it would be unsafe to call if the database is about to be closed, so there might be a race condition between calling `#close` and `#interrupt` at the same time. I'll add this to the docs. Anyways it is the responsibility of the programmer to use this responsibly.

Reply:
It turns out there is no race condition. I thought it does not count the state in between the _prepare and _step calls.